cl/lib/search_utils.py

```diff
@@ -586,6 +590,53 @@ def add_highlighting(
         main_params[f"f.{field}.hl.alternateField"] = field  # type: ignore
 
 
+def lookup_child_courts(parent_courts: list[str]) -> set[str]:
```

Comment:
I think your instinct to worry about performance was right, but I don't think your approach works very well. 

For example, checking the database cache for a key like `"child_courts:canb"` could well be slower than checking the Court table of the database for a key like `parent_court__id="canb"`. In the first case, you're checking a table with a lot of rows for a pretty large key. In the second case, you're checking a smaller table for a smaller key. I think the performance would be nearly identical. You could use the redis cache instead, but I bet your performance would be similar since Linux servers put commonly used data into memory, and the Court table is almost definitely in memory.

So...we should rethink the caching approach. 

But I think we need to rethink the query too, because our data is wide and shallow, not narrow and deep. In other words, our recursion will only ever be two or maybe three levels deep, but we could _easily_ have hundreds of courts to check for children. Your current code runs at least 1 query per court even if no courts have children. 

A better query would be something like (roughly): 

```
# Note this is parent_court_id, not parent_court__id. 
# I think that saves you a query too, but I'm not sure:
child_courts = Court.objects.filter(parent_court_id__in=parent_courts) 
```

That'd give you a list of all children for the queried courts. You can then recurse with the children you got, making a second query. If you then have grandchildren, you can recurse and make a third query, and so forth. I think this would bring your total query count to max of four, instead of potentially hundreds or even thousands.

As for caching, I think there are two approaches that are worth considering: 

1. The [`functools.cache`](https://docs.python.org/3/library/functools.html#functools.cache) would be nice. Alas, our pods are regularly reset, but this would be fastest, since it's store the results in Python's memory. Fun, but probably not the best approach until we make our pods more reliable.

2. Alphabetize the **complete** list of queried courts and make a single cache entry for the entire thing. 

Let me know what you think. Hopefully this will help.

Reply:
Thanks for the suggestions! I've applied them; the query you proposed is definitely a better approach. It reduces the queries significantly since it only requires a new query if at least one of the courts in the previous query has a child. And, since there aren't too many levels of children, the number of queries is just a few.

Regarding caching, I changed the approach to caching a group of courts instead of one by one, so this will also reduce the number of keys stored. First, I alphabetized the list of courts and then used them to create a key. I use hashing to generate the key so it has a fixed length, in case we get a huge query with many courts.

I think the only risk with this new caching approach is if we get an infinite number of combinations from users' queries. The cache table is going to grow a lot.